apsuite/commisslib/meas_touschek_lifetime.py

```diff
@@ -468,6 +389,301 @@ def plot_current_decay(self, fname=None, title=None):
+        # slc = slice(0, -window)
+        curr_a = curr_a[slc]
+        curr_b = curr_b[slc]
+        tim_a = tim_a[slc]
+        tim_b = tim_b[slc]
+        currt = currt[slc]
+        anly['current_a'] = curr_a
+        anly['current_b'] = curr_b
+        anly['tim_a'] = tim_a - tim_a[0]
+        anly['tim_b'] = tim_b - tim_a[0]
+        anly['current'] = currt
+
+        # Calc Touschek Lifetime from Total Lifetime of both bunches
+        if include_bunlen:
+            self._calc_touschek_lifetime()
+            tsck_a = anly['touschek_a']
+            tsck_b = anly['touschek_b']
+        else:
+            num = 1 - curr_b/curr_a
+            den = 1/ltime_a - 1/ltime_b
+            tsck_a = num/den
+            tsck_b = tsck_a * curr_a / curr_b
+            anly['touschek_a'] = tsck_a
+            anly['touschek_b'] = tsck_b
+
+        # Recover Gas Lifetime
+        gas_rate_a = 1/ltime_a - 1/tsck_a
+        gas_rate_b = 1/ltime_b - 1/tsck_b
+        anly['gas_lifetime_a'] = 1/gas_rate_a
+        anly['gas_lifetime_b'] = 1/gas_rate_b
+        anly['gas_lifetime'] = 2/(gas_rate_a + gas_rate_b)
+
+    def _calc_touschek_lifetime(self):
+        """."""
+        def model(curr, tous, blen):
+            curr_a, curr_b = curr
+            alpha_a = tous*curr_a/(1+blen*curr_a)
+            alpha_b = tous*curr_b/(1+blen*curr_b)
+            return alpha_a - alpha_b
+
+        anly = self.analysis
+        curr_a, curr_b = anly['current_a'], anly['current_b']
+        ltime_a, ltime_b = anly['total_lifetime_a'], anly['total_lifetime_b']
+        curr_a = curr_a[:ltime_a.size]
+        curr_b = curr_b[:ltime_b.size]
+
+        p0_ = (1, 1)
+        alpha_total = 1/ltime_a - 1/ltime_b
+        coeffs, pconv = _scy_opt.curve_fit(
+            model, (curr_a, curr_b), alpha_total, p0=p0_)
+        tous, blen = coeffs
+
+        alpha_a = tous*curr_a/(1+blen*curr_a)
+        alpha_b = tous*curr_b/(1+blen*curr_b)
+        anly['touschek_a'] = 1/alpha_a
+        anly['touschek_b'] = 1/alpha_b
+        anly['alpha_total'] = alpha_total
+        anly['tous_coeffs'] = coeffs
+        anly['tous_coeffs_pconv'] = pconv
+
+    def _do_measure(self):
+        meas = dict(
+            sum_a=[], sum_b=[], tim_a=[], tim_b=[], current=[],
+            rf_voltage=[], avg_pressure=[], tunex=[], tuney=[])
+        parms = self.params
+
+        curr = self.devices['currinfo']
+        rfcav = self.devices['rfcav']
+        tune = self.devices['tune']
+        press = self.pvs['avg_pressure']
+        bpm = self.devices[parms.bpm_name]
+        excx0 = self.pvs['exc_tunex_sts'].value
+        excy0 = self.pvs['exc_tuney_sts'].value
+
+        self.devices['trigger'].source = 'Study'
+        self.devices['event'].mode = 'Continuous'
+        self.devices['evg'].cmd_update_events()
+
+        self.turn_off_bpms_auto_monitor()
+        bpm.acq_nrsamples_pre = parms.acq_nrsamples_pre
+        bpm.acq_nrsamples_post = parms.acq_nrsamples_post
```

Comment:
I think we could set the property `bpm.acq_repeat` to `"normal"`. This way the BPM will not rearm the trigger right after the acquisition, which will allow us to set the mask without the need to `abort` the acquisition.